mne/preprocessing/_csd.py

```diff
@@ -0,0 +1,180 @@
+    sphere : tuple, (float, float, float, float)
+        The sphere, head-model of the form (x, y, z, r) where x, y, z
+        is the center of the sphere and r is the radius
+    n_jobs : int
+        The number of processes to run in parallel. Note. Only used for
+        Epochs input. Defaults to 1.
+    copy : bool
+        Whether to overwrite instance data or create a copy.
+
+    Returns
+    -------
+    inst_csd : instance of Epochs or Evoked
+        The transformed data. Output type will match input type.
+    """
+
+    _validate_type(inst, (BaseEpochs, BaseRaw, Evoked), 'inst')
+
+    if 'csd' in inst.info['comps']:
+        raise ValueError('CSD already applied, should not be reapplied')
+
+    if copy is True:
+        out = inst.copy()
+    else:
+        out = inst
+
+    if picks is None:
+        picks = pick_types(inst.info, meg=False, eeg=True, exclude='bads')
+    _validate_type(picks, (list, np.ndarray), 'picks')
+
+    if len(picks) == 0:
+        raise ValueError('No EEG channels found.')
+
+    if lambda2 is None:
+        lambda2 = 1e-5
+
+    _validate_type(lambda2, (float, int), 'lambda2')
+    if 0 > lambda2 or lambda2 > 1:
+        raise ValueError('lambda2 must be between 0 and 1, got %s' % lambda2)
+
+    _validate_type(stiffness, (float, int), 'stiffness')
+    if stiffness < 0:
+        raise ValueError('stiffness must be non-negative got %s' % stiffness)
+
+    _validate_type(n_legendre_terms, (int), 'n_legendre_terms')
+    if n_legendre_terms < 1:
+        raise ValueError('n_legendre_terms must be greater than 0, '
+                         'got %s' % n_legendre_terms)
+
+    _validate_type(sphere, tuple, 'sphere')
+    x, y, z, r = sphere
+    _validate_type(x, float, 'x')
+    _validate_type(y, float, 'y')
+    _validate_type(z, float, 'z')
+    _validate_type(r, float, 'r')
+    if r <= 0:
+        raise ValueError('sphere radius must be greater than 0, got %s' % r)
+
+    check_n_jobs(n_jobs)
+
+    _validate_type(copy, (bool), 'copy')
+
+    if ((g_matrix is None or h_matrix is None)):
+        pos = np.array([out.info['dig'][pick]['r'] for pick in picks])
+
+    G = _calc_g(np.dot(pos, pos.T), stiffness=stiffness,
+                num_lterms=n_legendre_terms)
+    H = _calc_h(np.dot(pos, pos.T), stiffness=stiffness,
+                num_lterms=n_legendre_terms)
+
+    G_precomputed = _prepare_G(G, lambda2)
+
+    if isinstance(out, BaseEpochs):
+        n_jobs = min(len(out), n_jobs)
+        logger.info('Using {} jobs'.format(n_jobs))
+        parallel, my_csd, _ = parallel_func(_compute_csd, n_jobs)
+        data = np.asarray(parallel(my_csd(e[picks],
+                                   G_precomputed=G_precomputed,
+                                   H=H, sphere=sphere) for e in out))
+        out.preload = True
+        out._data = data
+    else:
+        out._data = _compute_csd(out._data[picks], G_precomputed=G_precomputed,
+                                 H=H, sphere=sphere)
```

Comment:
Does the CSD end up being a linear transformation? If so then you can just apply it to `np.eye(n_eeg_ch)` exactly once, then `np.dot` this result with the data. No need for `n_jobs` or anything because it will always be fast.

Reply:
Yes, it's linear

Reply:
Great! Kill `n_jobs`, just apply it to `np.eye(n_eeg_chan)` to get the operator, then `np.dot` it with `raw._data`, `evoked._data`, or over each epoch in `epochs._data`

Reply:
Ok @drammock said there was a way to do Epochs with np.einsum. If anyone else wants to put it there, I would be interested to see. Basically you want to run 
for i in range(n_epochs):
    np.dot(trans_csd, inst._data[i])

where trans_csd is 6x6 for instance (n_channels x n_channels) and inst._data is (3, 6, 386) for instance (n_epochs, n_channels, n_times)

Reply:
You can but it will allocate an output memory block which you don't need to do. So in most cases (esp. bad memory limit cases) doing the iteration over epochs should be safer, and roughly as fast.

Reply:
Ok, I did it by reshaping and taking the dot product. If that's not good, I can change it.

Reply:
That will incur the same memory penalty as einsum but it will require the lines instead of one. I still prefer the iteration over epochs directly with no.dot in the loop